### PR TITLE
Add support for multiple database applications

### PR DESCRIPTION
**Pull request by sam-cw**
In ActiveRecord 6+ multiple databases can be used under specific roles. For example a read-replica can be used for read only roles. This requires a different database configuration structure and connection handling.

This PR allows Sequent to connect to databases configured in this way, though due to the nature of event sourcing it can only connect to a single database, which should always be the primary database.

### CODE REVIEW

File: lib/sequent/configuration.rb [hunk 1]
```diff
@@ -107,6 +109,13 @@ def initialize
 
       self.logger = Logger.new(STDOUT).tap { |l| l.level = Logger::INFO }
       self.error_locale_resolver = DEFAULT_ERROR_LOCALE_RESOLVER
+
+      self.enable_multiple_database_support = false
+      self.primary_database_key = :primary
+    end
+
+    def can_use_multiple_databases
```
Comment (lvonk) [inline lib/sequent/configuration.rb:117]: It would be nice to rename to `can_use_multiple_databases?`

File: lib/sequent/support/database.rb [hunk 2]
```diff
@@ -66,21 +101,27 @@ def self.with_schema_search_path(search_path, db_config, env = ENV['RACK_ENV'])
...
 
         if ActiveRecord::VERSION::MAJOR < 6
           ActiveRecord::Base.configurations[env.to_s] =
             ActiveSupport::HashWithIndifferentAccess.new(db_config).stringify_keys
         end
 
-        db_config[:schema_search_path] = search_path
-
-        ActiveRecord::Base.establish_connection db_config
+        if Sequent.configuration.can_use_multiple_databases
```
Comment (lvonk) [inline lib/sequent/support/database.rb:110]: Afterwards we should set the search_path back to what it was before like the previous version of this method. So in short:

```ruby
def self.can_use_multiple_databases(search_path)
  original_search_paths = get_current_search_path
  set_search_path(search_path)
  yield
ensure
  set_search_path(original_search_paths)
end
```

File: lib/sequent/support/database.rb [hunk 3]
```diff
@@ -30,17 +31,51 @@ def self.read_config(env)
       end
 
       def self.create!(db_config)
-        ActiveRecord::Base.establish_connection(db_config.merge(database: 'postgres'))
-        ActiveRecord::Base.connection.create_database(db_config[:database])
+        if Sequent.configuration.can_use_multiple_databases
+          ActiveRecord.legacy_connection_handling = false
```
Comment (lvonk) [inline lib/sequent/support/database.rb:35]: I see these couple of lines on multiple places, can we maybe  extract the into a single object?

**Review by lvonk — COMMENTED** [review]
Nice addition! Thanks a lot. Can we try to put the connection code into a single object?

File: lib/sequent/support/database.rb [hunk 4]
```diff
@@ -30,17 +31,51 @@ def self.read_config(env)
...
+        else
+          ActiveRecord::Base.establish_connection(db_config.merge(database: 'postgres'))
+          ActiveRecord::Base.connection.create_database(db_config[:database])
+        end
       end
 
       def self.drop!(db_config)
-        ActiveRecord::Base.establish_connection(db_config.merge(database: 'postgres'))
-        ActiveRecord::Base.connection.drop_database(db_config[:database])
+        if Sequent.configuration.can_use_multiple_databases?
```
Comment (bforma) [inline lib/sequent/support/database.rb:53]: This block of code looks similar to that in `create`. Can be dry it up into a separate method/class?

File: lib/sequent/support/database.rb [hunk 5]
```diff
@@ -66,21 +101,27 @@ def self.with_schema_search_path(search_path, db_config, env = ENV['RACK_ENV'])
...
         if ActiveRecord::VERSION::MAJOR < 6
           ActiveRecord::Base.configurations[env.to_s] =
             ActiveSupport::HashWithIndifferentAccess.new(db_config).stringify_keys
         end
 
-        db_config[:schema_search_path] = search_path
-
-        ActiveRecord::Base.establish_connection db_config
-
+        if Sequent.configuration.can_use_multiple_databases?
```
Comment (bforma) [inline lib/sequent/support/database.rb:110]: Idem.

**Review by bforma — APPROVED** [review]
Nice work!

**Pull request merged**